src/nbl/video/SOpenGLState.h

```diff
@@ -0,0 +1,173 @@
+#ifndef __NBL_S_OPENGL_STATE_H_INCLUDED__
+#define __NBL_S_OPENGL_STATE_H_INCLUDED__
+
+#include "nbl/core/IReferenceCounted.h"
+#include "COpenGLBuffer.h"
+#include "nbl/video/COpenGLRenderpassIndependentPipeline.h"
+#include "nbl/video/COpenGLPipelineLayout.h"
+#include "nbl/video/COpenGLDescriptorSet.h"
+#include "nbl/video/COpenGLComputePipeline.h"
+#include "nbl/video/IOpenGL_FunctionTable.h"
+#include "nbl/asset/ECommonEnums.h"
+#include <limits>
+
+namespace nbl {
+namespace video
+{
+
+struct SOpenGLState
+{
+    static inline constexpr uint32_t MAX_VIEWPORT_COUNT = 16u;
+
+    using buffer_binding_t = asset::SBufferBinding<const COpenGLBuffer>;
+
+    struct SVAO {
+        GLuint GLname;
+        uint64_t lastUsed;
+    };
+    struct HashVAOPair
+    {
+		COpenGLRenderpassIndependentPipeline::SVAOHash first = {};
+		SVAO second = { 0u,0ull };
+		//extra vao state being cached
+		std::array<buffer_binding_t, IGPUMeshBuffer::MAX_ATTR_BUF_BINDING_COUNT> vtxBindings;
+        buffer_binding_t idxBinding;
+        asset::E_INDEX_TYPE idxType; // not really state but i have to keep it somewhere
+
+        inline bool operator<(const HashVAOPair& rhs) const { return first < rhs.first; }
+    };
+    struct SDescSetBnd {
+        core::smart_refctd_ptr<const COpenGLPipelineLayout> pplnLayout;
+        core::smart_refctd_ptr<const COpenGLDescriptorSet> set;
+        core::smart_refctd_dynamic_array<uint32_t> dynamicOffsets;
+    };
+
+    using SGraphicsPipelineHash = std::array<GLuint, COpenGLRenderpassIndependentPipeline::SHADER_STAGE_COUNT>;
+
+    struct {
+        struct {
+            core::smart_refctd_ptr<const COpenGLRenderpassIndependentPipeline> pipeline;
+            SGraphicsPipelineHash usedShadersHash = { 0u, 0u, 0u, 0u, 0u };
+			GLuint usedPipeline = 0u;
+        } graphics;
+        struct {
+            core::smart_refctd_ptr<const COpenGLComputePipeline> pipeline;
+            GLuint usedShader = 0u;
+        } compute;
+    } pipeline;
+
+    struct {
+        core::smart_refctd_ptr<const COpenGLBuffer> buffer;
+    } dispatchIndirect;
+
+    struct {
+        // TODO add viewports to flushing routine
```

Comment:
just noting the TODO (btw this is also a dynamic state in Vulkan, optional, but we'll force it cause its stupid to put it in the pipeline)